storagemarket/lp2pimpl/net.go

```diff
@@ -167,7 +177,120 @@ func (p *DealProvider) handleNewAskStream(s network.Stream) {
+			dealState = &mktssm.ProviderDealState{
+				State:   mktssm.StorageDealUnknown,
+				Message: "deal not found: no deal with proposal cid " + q.Proposal.String(),
+			}
+		} else {
+			log.Errorw("failed to process deal status request", "err", err)
+			dealState = &mktssm.ProviderDealState{
+				State:   mktssm.StorageDealError,
+				Message: "server error: failed to process deal status request",
+			}
+		}
+	}
+
+	signature, err := p.sign(ctx, dealState)
+	if err != nil {
+		log.Errorw("failed to sign deal status response", "err", err)
+		return
+	}
+
+	response := &mktnet.DealStatusResponse{
+		DealState: *dealState,
+		Signature: *signature,
+	}
+
+	if err := cborutil.WriteCborRPC(s, response); err != nil {
+		log.Warnw("failed to write deal status response", "err", err)
+		return
+	}
+}
+
+func (p *DealProvider) getMarketsDealState(ctx context.Context, proposalCid cid.Cid) (*mktssm.ProviderDealState, error) {
+	deal, err := p.prov.DealBySignedProposalCID(ctx, proposalCid)
+	if err != nil {
+		return nil, err
+	}
+
+	signedPropNd, err := cborutil.AsIpld(&deal.ClientDealProposal)
+	if err != nil {
+		return nil, fmt.Errorf("calculating signed deal proposal ipld root: %w", err)
+	}
+	signedPropCid := signedPropNd.Cid()
+
+	dealState := mktssm.ProviderDealState{
+		State:         toLegacyMarketsState(deal),
+		Proposal:      &deal.ClientDealProposal.Proposal,
+		ProposalCid:   &signedPropCid,
+		PublishCid:    deal.PublishCID,
+		DealID:        deal.ChainDealID,
+		FastRetrieval: true,
+	}
+	return &dealState, nil
+}
+
+func (p *DealProvider) sign(ctx context.Context, dealState *mktssm.ProviderDealState) (*crypto.Signature, error) {
+	msg, err := cborutil.Dump(dealState)
+	if err != nil {
+		return nil, fmt.Errorf("sign deal: serializing deal state: %w", err)
+	}
+
+	minerAddr := p.prov.Address
+	mi, err := p.fullNode.StateMinerInfo(ctx, minerAddr, chaintypes.EmptyTSK)
+	if err != nil {
+		return nil, fmt.Errorf("sign deal: getting state miner info: %w", err)
+	}
+
+	signerAct, err := p.fullNode.StateAccountKey(ctx, mi.Worker, chaintypes.EmptyTSK)
+	if err != nil {
+		return nil, fmt.Errorf("sign deal: getting state account key: %w", err)
+	}
+
+	sig, err := p.fullNode.WalletSign(ctx, signerAct, msg)
+	if err != nil {
+		return nil, fmt.Errorf("sign deal: wallet sign: %w", err)
+	}
+	return sig, nil
+}
+
+func toLegacyMarketsState(deal *types.ProviderDealState) mktssm.StorageDealStatus {
+	if deal.Err != "" {
+		return mktssm.StorageDealError
+	}
+
+	switch deal.Checkpoint {
+	case dealcheckpoints.Accepted:
+		return mktssm.StorageDealProposalAccepted
+	case dealcheckpoints.Transferred:
+		return mktssm.StorageDealCheckForAcceptance
+	case dealcheckpoints.Published:
+		return mktssm.StorageDealPublishing
+	case dealcheckpoints.PublishConfirmed:
+		return mktssm.StorageDealStaged
+	case dealcheckpoints.AddedPiece:
+		return mktssm.StorageDealAwaitingPreCommit
+	case dealcheckpoints.Complete:
+		return mktssm.StorageDealActive
```

Comment:
We currently mark deals as complete only when they fail

Reply:
That case is taken care of on line 292:
```
	if deal.Err != "" {
		return mktssm.StorageDealError
	}
```